src/worker.rs

```diff
@@ -0,0 +1,320 @@
+        }
+    }
+
+    /// restart is a builder method that allows to modify the Restart settings
+    /// for the spawned worker
+    pub fn restart(mut self, restart: Restart) -> Self {
+        self.restart = restart;
+        self
+    }
+
+    /// shutdown is a builder method that allows to modify the Restart settings
+    /// for the spawned worker
+    pub fn shutdown(mut self, shutdown: Shutdown) -> Self {
+        self.shutdown = shutdown;
+        self
+    }
+
+    /// start spawns a new routine that executes the start function of this
+    /// Spec.
+    ///
+    /// ### Blocking on start function
+    ///
+    /// This function will block and wait until a notification from the
+    /// start function (executing on a newly spawned routine) is received; this
+    /// notification indicates if the start was successful or not.
+    ///
+    /// This blocking is necessary to ensure that supervised worker trees get
+    /// spawned in the specified/expected order.
+    pub async fn start(
+        mut self,
+        parent_ctx: &Context,
+        parent_name: &str,
+    ) -> anyhow::Result<Worker> {
+        let runtime_name = format!("{}/{}", parent_name, self.name);
+        let created_at = Utc::now();
+
+        // ON_START setup
+        let (started_tx, started_rx) = oneshot::channel::<Option<anyhow::Error>>();
+        let start_notifier = StartNotifier::from_oneshot(started_tx);
+
+        // CANCEL setup
+        let (ctx, termination_handle) = Context::with_cancel(parent_ctx);
+        let (kill_handle, kill_registration) = AbortHandle::new_pair();
+        let future = (*self.routine)(ctx, start_notifier);
+        let routine = Abortable::new(future, kill_registration);
+
+        // SPAWN -- actually spawn concurrent worker future
+        let join_handle = task::spawn(routine);
+
+        // ON_START blocking -- block before moving to the next start
+        let err = started_rx.await;
+
+        match err {
+            // On worker initialization failed, we need to signal
+            // this to starter
+            Err(err) => Err(anyhow::Error::new(err)),
+            Ok(Some(err)) => Err(err),
+
+            // Happy path
+            Ok(None) => Ok(Worker {
+                spec: self,
+                runtime_name,
+                created_at,
+                join_handle,
+                termination_handle,
+                kill_handle,
+            }),
+        }
+    }
+}
+
+impl Worker {
+    /// terminate tries to stop gracefuly a worker routine. In the scenario that
+    /// the routine doesn't stop after a timeout, it is brutally killed.
+    pub async fn terminate(self) -> Result<Spec, Arc<anyhow::Error>> {
+        self.termination_handle.abort();
+
+        let mut delay = time::delay_for(Duration::from_millis(1000));
+        tokio::select! {
+           _ = &mut delay => {
+               self.termination_handle.abort();
+               return Ok(self.spec)
+           },
```

Comment:
I think just using a `Timeout` for this and checking the returned Result for whether it timed out will be much simpler and accomplish the same effect.